Make RECIPES metadata drive

Jeff-Thompson12 opened the pull request.

Jeff-Thompson12 commented:
Hey team. This PR is going to fail the tests at the current moment because the file 'recipes.json' is in the main directory, but I would like some feedback regarding the direction I have taken on making the recipes metadata driven before I invest more time.

The parsing of the recipes file is done by S3 methods. When parsing the JSON file into R. Classes are assigned to either the blocks object and/or an individual block line. I think this approach will allow us to reach our main goal of allowing users to create their own tables and apply their own rules to inclusion and parsing, but would love to hear if you guys agree/disagree.

Jeff-Thompson12 commented:
I marked it ready for review, so at least one of you has to give me feedback now.

AARON-CLARK reviewed (changes requested):
Note to anyone testing this branch. Be sure to re-start your R session before hand!

@Jeff-Thompson12, overall, the architecture of step 1 looks really good! While reading, I often said "wow that's nice" to myself. lol 👍🏼 

One thing I'm noticing right away that when I select a recipe, and then select a second recipe, it's running the previous recipe's code again before running the newly selected recipe. I noticed this by just watching the console output (see image below). (1) corresponds to me clicking Recipe 19 and (2) is the output generated when I clicked on Recipe 18. Notice how (2) contains a bunch of joins by `AEBODSYS` and `AEDECOD` before moving on to the Recipe 19 joins ( on `AOCCFL`, `AESEV`, and `AESER`). Not a requirement, but perhaps we should add a message to the console that tells us when a recipe is clicked for more clarity.

![image](https://user-images.githubusercontent.com/17878953/221872636-b330a089-325c-471e-9d68-f0f8aeb42166.png)

Jeff-Thompson12 commented:
@AARON-CLARK We were experiencing a little bit of this before (mainly with the table title). Part of the problem is that the reactive graph for this module is a bit all over the place. The commit I just pushed seems to have corrected the problem for me. But users may have different experiences depending on the speed of their internet and computer. Not sure how to actually fix the problem without rethinking the reactive graph.

Jeff-Thompson12 commented:
@AARON-CLARK I cleaned up the reactive graph a little instead of using `debounce()`. One problem we were encountering is that the return of the JavaScript run for the recipes was causing the blocks and `input$recipe` to invalidate before `input$COLUMN`. (If you watch closely when you select the first recipe in the app as currently constructed, the first table will be built twice. First, without the group and secondly with the group.)

I also cleaned up the `pre_ADAE` and `pre_ADSL` reactives a little. There is still much work to be done there.

borgmaan commented on R/mod_tableGen_fct_filter_adae.R:
```diff
@@ -0,0 +1,173 @@
+filter_adae <- function(recipe, datalist, ADSL) {
```
A good example of a DSL like this in the wild can be found in the [CDISC Analysis Results Standard GitHub repo](https://github.com/cdisc-org/analysis-results-standard). [Here's an example](https://github.com/cdisc-org/analysis-results-standard/blob/b5d61bc29205805faf80ef3584dad7fb3fa472c7/workfiles/examples/PHUSE%20Connect%20Workshop/PHUSE%20Connect%20Examples.json#L89-L101) of their metadata syntax to define a common population filter:

```json
    {
      "id": "AnalysisSet_02_SAF",
      "condition": {
        "dataset": "ADSL",
        "variable": "SAFFL",
        "comparator": "EQ",
        "value": [
          "Y"
        ]
      },
      "order": 1,
      "label": "Safety Population"
    }
```

I think it will be good for us to reference CDISC ARS conventions as we go through the process of defining our metadata specification.

borgmaan reviewed (approved):
Hey -- sorry for the delay. I like the direction this is going. No changes needed for now. This is a good place to start with the initial specification and the base plumbing to use it throughout the application.

AARON-CLARK reviewed (commented):
Okay, so I test drove this and it looks like everything is 💪🏼. I'm even thinking about nominating it for the "PR of the week" award. I have a few comments:

(1) I did some time trials, and it looks like table 41 was taking 2min 23secs to render on `devel`. On shinyapps.io, this is a little faster at 50secs. This branch ran locally has it down 1min 16secs, so nearly cut in half! Great job! I assume this will be even faster on shinyapps.io. I had no idea this table was taking so long... when did this start happening? If things are taking this long, I think this necessitates the need for a progress bar. What do you think? Should we create an issue to add one?

(2) I personally like the format of the metadata layout because it is super intuitive given the layout of our dropzones. For fun, I tried to add a new table that looks like this:
```
"stan_0": {
    "title": "Table 0: A Fun Table Aaron like's making",
    "group_by": "TRT01P",
    "blocks": [{"data":"ADSL", "variable":"RANDFL", "statistic":"Y_FREQ"}, 
    {"data":"ADSL", "variable":"SAFFL", "statistic":"FREQ"}, 
    {"data":"ADSL", "variable":"SEX", "statistic":"FREQ"},
    {"data":"ADSL", "variable":"ETHNIC", "statistic":"FREQ"},
    {"data":"ADSL", "variable":"RACE", "statistic":"FREQ"}]
  }
```
It's basically just a mash up of table 3 and 5. It worked great! But what if I wanted to add a population filter to the table. I see that process is completely separate from the metadata. Is there any way we could somehow include the population config in the `recipes.json` file so users (like me) only have to update info in one location? What do you think about @borgmaan comment from a few weeks ago about following a ADR format?


(3) I left a comment below that asks about the `stat_options_fn` parameter in the metadata. I realize it somehow controls the label that's applied in the table, but boy, its been a minute since I've looked at this code. All the more reason that this PR desperately needs a vignette paired with it to describe how to use these parameters to add/ change / remove metadata recipes and how each change will impact the output.

AARON-CLARK commented:
> If things are taking this long, I think this necessitates the need for a progress bar. What do you think? Should we create an issue to add one?

@Jeff-Thompson12, what do you think about the prospect of adding a progress bar the the table generator? Obviously in a different PR.

AARON-CLARK commented:
> This is where we are heading. This PR is just step one. I am just setting up the framework to allow us to accomplish this goal.

@Jeff-Thompson12, sounds good. I'll create a specific issue to address this and put it on our roadmap.

Jeff-Thompson12 commented:
> @Jeff-Thompson12, what do you think about the prospect of adding a progress bar the the table generator? Obviously in a different PR.

I think we might want to go even farther and think about when table generation should be triggered and what UI elements we want the user to be able to interact with while a table is being generated.

AARON-CLARK commented on vignettes/dev04_RECIPES.Rmd:
```diff
@@ -0,0 +1,61 @@
+
+### The `blocks` object
+
+There is one additional sub-object not yet mentioned, the `recipe_inclusion` object, which can assign a class to the `blocks` object. The standard inclusion criteria is for all data sets in a `blocks` object to be present in the list of data sets uploaded and for all variables listed in the `block` objects to be present in the data set identified with them. Assigning a value to `recipe_inclusion` ...
```
I think this paragraph deserves an example. I saw below that it's associated with table 41, but I'm still not sure I understand what it does. What's an example for your example on `crit_01`? lol

AARON-CLARK commented on vignettes/dev04_RECIPES.Rmd:
```diff
@@ -0,0 +1,61 @@
+
+### The `block` object
+
+Currently only one potential class is being assigned to the `block` object and it is the value of element `stat_options_fn`. For certain statistics, such as `MEAN`, a drop down is provided in the statistic block (e.g. `AVISIT` values for `MEAN`). In the event the user wants a block to output multiple weeks, they can provide a value for `stat_options_fn` and its corresponding function (i.e. A valu ...
```
```suggestion
Currently, only one potential class is being assigned to the `block` object and it is the value of element `stat_options_fn`. If you recall, when the `MEAN` statistic is selected in the app, a dropdown list is displayed containing different `AVISIT` values. Thus, if the user wants a block to output multiple weeks, they can provide those values to `stat_options_fn`. Its corresponding function ( for example, with value of `Baseline` would correspond to `stat_options.Baseline()`). The user could also opt to hard code these values using the element `stat_options`.
```

AARON-CLARK replied:
I tried to provide a more concrete example instead of using the `_01` method. How did I do?

AARON-CLARK commented on vignettes/dev04_RECIPES.Rmd:
```diff
@@ -0,0 +1,61 @@
+
+Currently only one potential class is being assigned to the `block` object and it is the value of element `stat_options_fn`. For certain statistics, such as `MEAN`, a drop down is provided in the statistic block (e.g. `AVISIT` values for `MEAN`). In the event the user wants a block to output multiple weeks, they can provide a value for `stat_options_fn` and its corresponding function (i.e. A valu ...
+
+The element `stat_arg` is used to determine the input for a statistic with a drop down. These two elements together can allow the user to create multiple generated blocks with only one `block` object.
```
```suggestion
Similarly, the element `stat_arg` is used to determine the input for a statistic with a drop down. Paired with `stat_options_fn`, these two elements together can allow the user to create multiple generated blocks with only one `block` object.
```

AARON-CLARK reviewed (changes requested):
Couple more suggestions.

Jeff-Thompson12 commented:
@AARON-CLARK I more or less took your edits. I have also added in some code chunks to flesh out the "Parsing" section of the vignette.
